pkg/migrate/cstor/volume.go

```diff
@@ -0,0 +1,902 @@
+			return errors.Wrapf(err, "failed to migrate pvc to csi spec")
+		}
+	} else {
+		klog.Infof("PVC and storageclass already migrated to csi format")
+	}
+	v.StorageClass, err = v.KubeClientset.StorageV1().
+		StorageClasses().Get(*pvcObj.Spec.StorageClassName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	_, err = v.migratePV(pvcObj)
+	if err != nil {
+		return errors.Wrapf(err, "failed to migrate pv to csi spec")
+	}
+	err = v.removeOldTarget()
+	if err != nil {
+		return errors.Wrapf(err, "failed to remove old target deployment")
+	}
+	klog.Infof("Creating CVC to bound the volume and trigger CSI driver")
+	err = v.createCVC(v.PVName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create cvc")
+	}
+	err = v.validateMigratedVolume()
+	if err != nil {
+		return errors.Wrapf(err, "failed to validate migrated volume")
+	}
+	err = v.patchTargetPodAffinity()
+	if err != nil {
+		return errors.Wrap(err, "failed to patch target affinity")
+	}
+	err = v.cleanupOldResources()
+	if err != nil {
+		return errors.Wrapf(err, "failed to cleanup old volume resources")
+	}
+	return nil
+}
+
+func (v *VolumeMigrator) migratePVC(pvObj *corev1.PersistentVolume) (*corev1.PersistentVolumeClaim, error) {
+	klog.Infof("Generating equivalent CSI PVC")
+	pvcObj, recreateRequired, err := v.generateCSIPVC(pvObj.Name)
+	if err != nil {
+		return nil, err
+	}
+	if recreateRequired {
+		klog.Infof("Recreating equivalent CSI PVC")
+		pvcObj, err = v.RecreatePVC(pvcObj)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return pvcObj, nil
+}
+
+func (v *VolumeMigrator) migratePV(pvcObj *corev1.PersistentVolumeClaim) (*corev1.PersistentVolume, error) {
+	klog.Infof("Generating equivalent CSI PV")
+	pvObj, recreateRequired, err := v.generateCSIPV(pvcObj.Spec.VolumeName, pvcObj)
+	if err != nil {
+		return nil, err
+	}
+	if recreateRequired {
+		klog.Infof("Recreating equivalent CSI PV")
+		_, err = v.RecreatePV(pvObj)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return pvObj, nil
+}
+
+func (v *VolumeMigrator) generateCSIPVC(pvName string) (*corev1.PersistentVolumeClaim, bool, error) {
+	pvObj, err := v.KubeClientset.CoreV1().
+		PersistentVolumes().
+		Get(pvName, metav1.GetOptions{})
+	if err != nil {
+		return nil, false, err
+	}
+	pvcName := pvObj.Spec.ClaimRef.Name
+	pvcNamespace := pvObj.Spec.ClaimRef.Namespace
+	pvcObj, err := v.KubeClientset.CoreV1().PersistentVolumeClaims(pvcNamespace).
+		Get(pvcName, metav1.GetOptions{})
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return nil, false, err
+		}
+	}
+	if pvcObj.Annotations["volume.beta.kubernetes.io/storage-provisioner"] != cstorCSIDriver {
+		csiPVC := &corev1.PersistentVolumeClaim{}
+		csiPVC.Name = pvcName
+		csiPVC.Namespace = pvcNamespace
+		csiPVC.Annotations = map[string]string{
+			"volume.beta.kubernetes.io/storage-provisioner": cstorCSIDriver,
+		}
+		csiPVC.Spec.AccessModes = pvObj.Spec.AccessModes
+		csiPVC.Spec.Resources.Requests = pvObj.Spec.Capacity
```

Comment:
Get capacity from PV object so if any user performed resize for previous volumes it will be up to date(getting from CStorVolume will give more confidence).

Reply:
Here you mean: `Get capacity from PV object` right?

Reply:
We can get it CV if it is possible (or) else PV is also fine(I worried about the users who skipped last step execution).